solr/core/src/java/org/apache/solr/response/transform/ChildDocTransformer.java

```diff
@@ -0,0 +1,257 @@
+import org.apache.lucene.search.join.BitSetProducer;
+import org.apache.lucene.util.BitSet;
+import org.apache.solr.common.SolrDocument;
+import org.apache.solr.response.DocsStreamer;
+import org.apache.solr.schema.IndexSchema;
+import org.apache.solr.search.DocSet;
+import org.apache.solr.search.SolrDocumentFetcher;
+import org.apache.solr.search.SolrIndexSearcher;
+import org.apache.solr.search.SolrReturnFields;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.apache.solr.response.transform.ChildDocTransformerFactory.NUM_SEP_CHAR;
+import static org.apache.solr.response.transform.ChildDocTransformerFactory.PATH_SEP_CHAR;
+import static org.apache.solr.schema.IndexSchema.NEST_PATH_FIELD_NAME;
+
+class ChildDocTransformer extends DocTransformer {
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+  private static final String ANON_CHILD_KEY = "_childDocuments_";
+
+  private final String name;
+  private final BitSetProducer parentsFilter;
+  private final DocSet childDocSet;
+  private final int limit;
+
+  private final SolrReturnFields childReturnFields = new SolrReturnFields();
+
+  ChildDocTransformer(String name, BitSetProducer parentsFilter,
+                      DocSet childDocSet, int limit) {
+    this.name = name;
+    this.parentsFilter = parentsFilter;
+    this.childDocSet = childDocSet;
+    this.limit = limit;
+  }
+
+  @Override
+  public String getName()  {
+    return name;
+  }
+
+  @Override
+  public void transform(SolrDocument rootDoc, int rootDocId) {
+    // note: this algorithm works if both if we have have _nest_path_  and also if we don't!
+
+    try {
+
+      // lookup what the *previous* rootDocId is, and figure which segment this is
+      final SolrIndexSearcher searcher = context.getSearcher();
+      final List<LeafReaderContext> leaves = searcher.getIndexReader().leaves();
+      final int seg = ReaderUtil.subIndex(rootDocId, leaves);
+      final LeafReaderContext leafReaderContext = leaves.get(seg);
+      final int segBaseId = leafReaderContext.docBase;
+      final int segRootId = rootDocId - segBaseId;
+      final BitSet segParentsBitSet = parentsFilter.getBitSet(leafReaderContext);
+      final int segPrevRootId = segRootId==0? -1: segParentsBitSet.prevSetBit(segRootId - 1); // can return -1 and that's okay
+
+      if(segPrevRootId == (segRootId - 1)) {
+        // doc has no children, return fast
+        return;
+      }
+
+      // we'll need this soon...
+      final SortedDocValues segPathDocValues = DocValues.getSorted(leafReaderContext.reader(), NEST_PATH_FIELD_NAME);
+
+      // the key in the Map is the document's ancestors key(one above the parent), while the key in the intermediate
+      // MultiMap is the direct child document's key(of the parent document)
+      Map<String, Multimap<String, SolrDocument>> pendingParentPathsToChildren = new HashMap<>();
+
+      IndexSchema schema = searcher.getSchema();
```

Comment:
Master has changes recently committed related to document fetching.  It'll simplify this nicely here.  Can you please sync up?

Reply:
Will do